webapp/src/ts/services/fast-action-button.service.ts

```diff
@@ -0,0 +1,236 @@
+        },
+      }))
+      .sort((a, b) => a.label?.localeCompare(b.label));
+  }
+
+  private getContactFormActions(childContactTypes, parentFacilityId, queryParams?): FastAction[] {
+    return (childContactTypes || [])
+      .map(contactType => ({
+        id: contactType.id,
+        label: this.getFormTitle(contactType.create_key) || contactType.id,
+        icon: { name: contactType.icon, type: IconType.RESOURCE },
+        canDisplay: () => this.authService.has(['can_edit', 'can_create_places']),
+        execute: () => {
+          const route = ['/contacts', 'add', contactType.id];
+          if (parentFacilityId) {
+            // Inserting facility's ID between "/contacts" and "add" router segments.
+            route.splice(1, 0, parentFacilityId);
+          }
+          this.router.navigate(route, { queryParams });
+        },
+      }))
+      .sort((a, b) => a.label?.localeCompare(b.label));
+  }
+
+  private getSendMessageAction(context: CommunicationActionsContext, config: CommunicationActionsConfig = {}) {
+    const { sendTo, callbackOpenSendMessage } = context;
+    const { isPhoneRequired, useMailtoInMobile } = config;
+
+    const validatePhone = () => isPhoneRequired ? !!sendTo?.phone : true;
+    const canUseMailto = () => useMailtoInMobile && this.responsiveService.isMobile();
+
+    return {
+      id: 'send-message',
+      labelKey: 'fast_action_button.send_message',
+      icon: { name: 'fa-envelope', type: IconType.FONT_AWESOME },
+      canDisplay: async () => {
+        const permission = [ 'can_view_message_action' ];
+        !canUseMailto() && permission.push('can_edit');
+        return validatePhone() && await this.authService.has(permission);
+      },
+      execute: () => {
+        if (canUseMailto()) {
+          this.executeMailto(`sms:${sendTo?.phone}`);
+          return;
+        }
+        callbackOpenSendMessage && callbackOpenSendMessage(sendTo?._id);
+      },
+    };
+  }
+
+  private getPhoneAction(context: CommunicationActionsContext) {
+    return {
+      id: 'phone-call',
+      labelKey: 'fast_action_button.phone_call',
+      icon: { name: 'fa-phone', type: IconType.FONT_AWESOME },
+      canDisplay: async () => {
+        return context.sendTo?.phone
+          && await this.authService.has('can_view_call_action')
+          && this.responsiveService.isMobile();
+      },
+      execute: () => this.executeMailto(`tel:${context.sendTo?.phone}`),
+    };
+  }
+
+  private getUpdateFacilityAction(reportContentType) {
+    return {
+      id: 'update-facility',
+      labelKey: 'fast_action_button.update_facility',
+      icon: { name: 'fa-pencil', type: IconType.FONT_AWESOME },
+      canDisplay: async () => reportContentType !== 'xml' && await this.authService.has('can_edit'),
+      execute: () => this.reportsActions.launchEditFacilityDialog(),
+    };
+  }
+
+  getReportRightSideActions(context: ReportActionsContext): Promise<FastAction[]> {
```

Comment:
all of these `get*Actions` functions are sync so we can type the return type as just `FastAction[]`
```suggestion
  getReportRightSideActions(context: ReportActionsContext): FastAction[] {
```

Reply:
They have `return this.filterActions(actions);` and that function awaits for a code [here](https://github.com/medic/cht-core/blob/0591dd86828d7d24af3666f77b8cc907f79544dc/webapp/src/ts/services/fast-action-button.service.ts#L43). At the end it's a promise of FastActions

Reply:
oops, good catch! I hadn't noticed `filterActions` was async too